File: app/src/main/java/br/com/valenstech/letraviva/ui/login/LoginActivity.java
package br.com.valenstech.letraviva.ui.login;

import android.app.Activity;
import android.content.Intent;
import android.os.Bundle;

import androidx.activity.result.ActivityResultLauncher;
import androidx.activity.result.contract.ActivityResultContracts;
import androidx.annotation.Nullable;
import androidx.appcompat.app.AppCompatActivity;
import androidx.lifecycle.ViewModelProvider;

import com.google.android.gms.auth.api.signin.GoogleSignIn;
import com.google.android.gms.auth.api.signin.GoogleSignInAccount;
import com.google.android.gms.auth.api.signin.GoogleSignInClient;
import com.google.android.gms.auth.api.signin.GoogleSignInOptions;
import com.google.android.gms.common.api.ApiException;

import br.com.valenstech.letraviva.R;
import br.com.valenstech.letraviva.databinding.ActivityLoginBinding;
import br.com.valenstech.letraviva.ui.home.MainActivity;
import br.com.valenstech.letraviva.viewmodel.AutenticacaoViewModel;

public class LoginActivity extends AppCompatActivity {

    private ActivityLoginBinding binding;
    private AutenticacaoViewModel autenticacaoViewModel;
    private GoogleSignInClient clienteGoogle;

    private final ActivityResultLauncher<Intent> resultadoLogin =
            registerForActivityResult(new ActivityResultContracts.StartActivityForResult(), resultado -> {
                if (resultado.getResultCode() != Activity.RESULT_OK) {
                    return;
                }
                Intent dados = resultado.getData();
                if (dados == null) {
                    return;
                }
                try {
                    GoogleSignInAccount conta = GoogleSignIn.getSignedInAccountFromIntent(dados)
                            .getResult(ApiException.class);
                    autenticarComGoogle(conta);
                } catch (ApiException excecao) {
                    // falha de autenticação ignorada para manter experiência
                }
            });

    @Override
    protected void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        binding = ActivityLoginBinding.inflate(getLayoutInflater());
        setContentView(binding.getRoot());

        autenticacaoViewModel = new ViewModelProvider(this).get(AutenticacaoViewModel.class);
        autenticacaoViewModel.getUsuario().observe(this, usuario -> {
            if (usuario != null) {
                iniciarPrincipal();
            }
        });
        autenticacaoViewModel.verificarUsuario();

        GoogleSignInOptions opcoesGoogle = new GoogleSignInOptions.Builder(GoogleSignInOptions.DEFAULT_SIGN_IN)
                .requestIdToken(getString(R.string.default_web_client_id))
                .requestEmail()
                .build();
        clienteGoogle = GoogleSignIn.getClient(this, opcoesGoogle);

        binding.btnGoogle.setOnClickListener(view -> {
            Intent intencaoLogin = clienteGoogle.getSignInIntent();
            resultadoLogin.launch(intencaoLogin);
        });
    }

    private void autenticarComGoogle(@Nullable GoogleSignInAccount conta) {
        if (conta == null) {
            return;
        }
        String token = conta.getIdToken();
        if (token == null) {
            return;
        }
        autenticacaoViewModel.autenticarComGoogle(token);
    }

    private void iniciarPrincipal() {
        Intent intent = new Intent(this, MainActivity.class);
        startActivity(intent);
        finish();
    }
}

File: app/src/main/java/br/com/valenstech/letraviva/ui/login/LoginActivity.kt
package br.com.valenstech.letraviva.ui.login

import android.app.Activity
import android.content.Intent
import android.os.Bundle
import androidx.activity.result.contract.ActivityResultContracts
import androidx.appcompat.app.AppCompatActivity
import androidx.lifecycle.ViewModelProvider
import br.com.valenstech.letraviva.databinding.ActivityLoginBinding
import br.com.valenstech.letraviva.ui.home.MainActivity
import br.com.valenstech.letraviva.viewmodel.AuthViewModel
import com.google.android.gms.auth.api.signin.GoogleSignIn
import com.google.android.gms.auth.api.signin.GoogleSignInAccount
import com.google.android.gms.auth.api.signin.GoogleSignInClient
import com.google.android.gms.auth.api.signin.GoogleSignInOptions
import com.google.android.gms.common.api.ApiException
import com.google.firebase.auth.FirebaseAuth

class LoginActivity : AppCompatActivity() {

    private lateinit var binding: ActivityLoginBinding
    private lateinit var authViewModel: AuthViewModel
    private lateinit var googleSignInClient: GoogleSignInClient

    private val signInLauncher = registerForActivityResult(
        ActivityResultContracts.StartActivityForResult()
    ) { result ->
        if (result.resultCode == Activity.RESULT_OK) {
            val task = GoogleSignIn.getSignedInAccountFromIntent(result.data)
            try {
                val account = task.getResult(ApiException::class.java)
                firebaseAuthWithGoogle(account)
            } catch (e: ApiException) {
                // handle
            }
        }
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        binding = ActivityLoginBinding.inflate(layoutInflater)
        setContentView(binding.root)

        authViewModel = ViewModelProvider(this)[AuthViewModel::class.java]
        authViewModel.user.observe(this) { user ->
            if (user != null) {
                startActivity(Intent(this, MainActivity::class.java))
                finish()
            }
        }
        authViewModel.checkUser()

        val gso = GoogleSignInOptions.Builder(GoogleSignInOptions.DEFAULT_SIGN_IN)
            .requestIdToken(getString(br.com.valenstech.letraviva.R.string.default_web_client_id))
            .requestEmail()
            .build()
        googleSignInClient = GoogleSignIn.getClient(this, gso)

        binding.btnGoogle.setOnClickListener {
            val signInIntent = googleSignInClient.signInIntent
            signInLauncher.launch(signInIntent)
        }
    }

    private fun firebaseAuthWithGoogle(account: GoogleSignInAccount?) {
        account?.idToken?.let { token ->
            authViewModel.firebaseAuthWithGoogle(token)
        }
    }
}
